server/src/main/java/org/apache/druid/server/coordination/SegmentLoadDropHandler.java

```diff
@@ -90,19 +93,46 @@ public class SegmentLoadDropHandler implements DataSegmentChangeHandler
   private final SegmentManager segmentManager;
   private final ScheduledExecutorService exec;
   private final ServerTypeConfig serverTypeConfig;
-  private final ConcurrentSkipListSet<DataSegment> segmentsToDelete;
+  private final ConcurrentSkipListSet<DataSegment> segmentsToDrop;
   private final SegmentCacheManager segmentCacheManager;
 
   private volatile boolean started = false;
 
-  // Keep history of load/drop request status in a LRU cache to maintain idempotency if same request shows up
-  // again and to return status of a completed request. Maximum size of this cache must be significantly greater
-  // than number of pending load/drop requests. so that history is not lost too quickly.
-  private final Cache<DataSegmentChangeRequest, AtomicReference<Status>> requestStatuses;
+  /**
+   * Used to cache the status of a completed load or drop request until it has
+   * been served to the (Coordinator) client exactly once.
+   * <p>
+   * The cache is used as follows:
+   * <ol>
+   * <li>An entry with state PENDING is added to the cache upon receiving a
+   * request to load or drop a segment.</li>
+   * <li>A duplicate request received at this point is immediately answered with PENDING.</li>
+   * <li>Once the load/drop finishes, the entry is updated to either SUCCESS or FAILED.</li>
+   * <li>A duplicate request received at this point is immediately answered with
+   * SUCCESS or FAILED and the entry is removed from the cache.</li>
+   * <li>If the first request itself finishes after the load or drop has already
+   * completed, it is answered with a SUCCESS or FAILED and the entry is removed
+   * from the cache.</li>
+   * </ol>
+   * <p>
+   * Maximum size of this cache must be significantly greater than the number of
```

Comment:
If the Coordinator sends a batch larger than the cache size, the Historical would not be able to have an entry for some of the requests in the cache. These requests would still be processed but their finished statuses would never be updated in the cache and thus never sent back to the Coordinator. 

The Coordinator would end up retrying these requests and the Historical might duplicate some work but there would never be any correctness issue, since we are not sending any response to the Coordinator for the un-cached requests.

Given that the default cache size is 100 and default request batch size is 1, we should be okay. We can update the docs to mention that the cache size must be at least 10x of the batch size.

Reply:
Interestingly, `druid.segmentCache.statusQueueMaxSize` is not documented yet.

Reply:
Where does the 10x factor come from?

Reply:
> Where does the 10x factor come from?

Just a safe-side measure in case Coordinator sends multiple batches of different requests. Not really based on any concrete thoughts.

Reply:
Some more thoughts on this:
- The `Cache` can really just be a `Map`. The only advantage the `Cache` might offer is auto-expiry of requests that the Coordinator gave up on, but I don't think that is currently being used in the code anyway.
- Maybe we shouldn't even have a config for `batchSize`. The Coordinator would start by sending a single request to the historicals and in each HTTP response, each Historical could indicate how many requests it would like to take in the next batch. This could be based on the number of loading threads, number of pending requests.
- Given that the number of pending requests is always going to be small, we should just get rid of the `cacheSize` config and hard-code it to something like 1000. And if in any case, the Coordinator sends more requests than a Historical can handle, it immediately fails the request with an appropriate error message.

What do you think?

Reply:
> The `Cache` can really just be a `Map`. The only advantage the `Cache` might offer is auto-expiry of requests that the Coordinator gave up on, but I don't think that is currently being used in the code anyway.

The Coordinator could crash and never follow up on a request, so it's still useful to expire stuff.

> Maybe we shouldn't even have a config for batchSize. The Coordinator would start by sending a single request to the historicals and in each HTTP response, each Historical could indicate how many requests it would like to take in the next batch. This could be based on the number of loading threads, number of pending requests.

Seems like a good idea! In addition, in case the Coordinator sends more commands than the Historical asks for, the Historical could reject the extra commands. (Although I am not sure if the protocol allows this.)

> Given that the number of pending requests is always going to be small, we should just get rid of the cacheSize config and hard-code it to something like 1000.

Sounds good to me. Fewer configs are good.

Btw, I am wondering now, what are the scenarios where the Historical might receive multiple copies of the same load or drop command? Is this something that happens in typical operation, or in some kind of abnormal case?

Reply:
> Btw, I am wondering now, what are the scenarios where the Historical might receive multiple copies of the same load or drop command?

Do you mean the same action on same segment or different actions on same segment? Both of these can actually happen under normal conditions:

#### Same action on same segment
Even under normal operation, the Coordinator keeps sending the same request to a Historical until the request finishes.
Essentially, the Coordinator sends a batch of requests to the Historical and waits for a response. Historical sends back a response as soon as any _one_ of the requests in this batch (or even a previously submitted batch which is still being tracked) finishes. So chances are that the Coordinator gets back a response with one request having status SUCCESS or FAILED and the others PENDING, because the Historical was still working on them when it completed the HTTP request. The Coordinator then re-sends the requests for which it had received a PENDING status. By then, the Historical might have already finished them and cached their results.

#### Different actions on same segment but after a delay
The case for a different action (i.e. load after drop) without ever having read the status of the previous action can happen during coordinator restarts or leader re-elections. Although, in these cases, it is not likely that the drop and load would happen concurrently.

#### Different actions on same segment happening concurrently
This is not expected to happen under normal conditions, and can only be caused by the Coordinator being in some weird state or a bug in the code.